File: golang/dynamic_programming/longest_valid_parent.go
package main

import "fmt"

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func longestValidParentheses(s string) int {
	if len(s) < 2 {
		return 0
	}

	var result int
	var count int
	longest := make([]int, len(s))
	for i := 0; i < len(s); i++ {
		if s[i] == '(' {
			count++
		} else if count > 0 {
			longest[i] = longest[i-1] + 2
			if i-longest[i] >= 0 {
				longest[i] += longest[i-longest[i]]
			}
			result = max(result, longest[i])
		}
	}
	return result
}

func main() {
	fmt.Println(longestValidParentheses("()"))
}

File: rust/longest_valid_parent.rs
use std::cmp;
fn longest_valid_parentheses(s: String) -> i32 {
    if s.len() < 2 {
        return 0;
    }
    let mut result = 0;
    let mut longest = vec![0; s.len()];
    let mut count = 0;

    for (i, c) in s.chars().enumerate() {
        if c == '(' {
            count += 1;
        } else if count > 0 {
            longest[i] = longest[i - 1] + 2;
            if i > longest[i] {
                longest[i] += longest[i - longest[i]]
            }
            result = cmp::max(result, longest[i]);
            count -= 1;
        }
    }
    result as i32
}

fn main() {
    let s = String::from("()");
    println!("the result is {}", longest_valid_parentheses(s))
}
